cmd/soroban-rpc/internal/db/event.go

```diff
@@ -97,3 +164,185 @@ func (eventHandler *eventHandler) InsertEvents(lcm xdr.LedgerCloseMeta) error {
 
 	return nil
 }
+
+type ScanFunction func(
+	event xdr.DiagnosticEvent,
+	cursor Cursor,
+	ledgerCloseTimestamp int64,
+	txHash *xdr.Hash,
+) bool
+
+// trimEvents removes all Events which fall outside the ledger retention window.
+func (eventHandler *eventHandler) trimEvents(latestLedgerSeq uint32, retentionWindow uint32) error {
+	if latestLedgerSeq+1 <= retentionWindow {
+		return nil
+	}
+	cutoff := latestLedgerSeq + 1 - retentionWindow
+	id := Cursor{Ledger: cutoff}.String()
+
+	_, err := sq.StatementBuilder.
+		RunWith(eventHandler.stmtCache).
+		Delete(eventTableName).
+		Where(sq.Lt{"id": id}).
+		Exec()
+	return err
+}
+
+// GetEvents applies f on all the events occurring in the given range with specified contract IDs if provided.
+// The events are returned in sorted ascending Cursor order.
+// If f returns false, the scan terminates early (f will not be applied on
+// remaining events in the range).
+//
+//nolint:funlen
+func (eventHandler *eventHandler) GetEvents(
+	ctx context.Context,
+	cursorRange CursorRange,
+	contractIDs [][]byte,
+	topics [][]string,
+	eventTypes []int,
+	f ScanFunction,
+) error {
+	start := time.Now()
+
+	rowQ := sq.
+		Select(" id", "event_data", "transaction_hash", "ledger_close_time").
+		From(eventTableName).
+		Where(sq.GtOrEq{"id": cursorRange.Start.String()}).
+		Where(sq.Lt{"id": cursorRange.End.String()}).
+		OrderBy("id ASC")
+
+	if len(contractIDs) > 0 {
+		rowQ = rowQ.Where(sq.Eq{"contract_id": contractIDs})
+	}
+	if len(eventTypes) > 0 {
+		rowQ = rowQ.Where(sq.Eq{"event_type": eventTypes})
+	}
+
+	if len(topics) > 0 {
+		var orConditions sq.Or
+		for i, topic := range topics {
+			if topic == nil {
+				continue
+			}
+			orConditions = append(orConditions, sq.Eq{fmt.Sprintf("topic%d", i+1): topic})
+		}
+		if len(orConditions) > 0 {
+			rowQ = rowQ.Where(orConditions)
+		}
+	}
+
+	rows, err := eventHandler.db.Query(ctx, rowQ)
+	if err != nil {
+		return fmt.Errorf(
+			"db read failed for start ledger cursor= %v end ledger cursor= %v "+
+				"contractIDs= %v eventTypes= %v topics= %v   : %w",
+			cursorRange.Start.String(),
+			cursorRange.End.String(),
+			contractIDs,
+			eventTypes,
+			topics,
+			err)
+	}
+
+	defer rows.Close()
+
+	foundRows := false
+	for rows.Next() {
```

Comment:
You also need to check `rows.Err()` somewhere, I think? I forget the exact semantics